JSON-Grammar-spec.md

```diff
@@ -0,0 +1,794 @@
+
+#### AST Transformations
+- **`ast`**: Optional object defining custom AST generation rules
+- Maps rule names to JSON expressions for AST transformation
+- Overrides default AST generation behavior
+
+## Grammar Node Types
+
+The JSON grammar specification supports five fundamental node types for defining parsing rules. Each type has a specific JSON representation with both full interface and shorthand syntax options.
+
+### 1. RefNode (Reference Node)
+
+References a named rule defined elsewhere in the grammar.
+
+#### Interface
+```typescript
+type RefNode<Name extends string = string> = {
+  r: Name;
+};
+```
+
+#### Syntax
+```json
+{"r": "RuleName"}
+```
+
+#### Examples
+```json
+{
+  "start": "Program",
+  "cst": {
+    "Program": {"r": "Statement"},
+    "Statement": "return;"
+  }
+}
+```
+
+### 2. TerminalNode (Terminal Node)
+
+Matches literal strings, regular expressions, or arrays of alternatives. Terminal nodes are leaf nodes in the parse tree.
+
+#### Interface
+```typescript
+interface TerminalNode {
+  type?: string;                           // Type name (default: "Text")
+  t: RegExp | string | '' | string[];      // Pattern(s) to match
+  repeat?: '*' | '+';                      // Repetition (only for string arrays)
+  sample?: string;                         // Sample text for generation
+  ast?: AstNodeExpression;                 // AST transformation
+}
+
+// Shorthand: string, RegExp, or empty string
+type TerminalNodeShorthand = RegExp | string | '';
+```
+
+#### Syntax Options
+
+**String Literal:**
+```json
+"hello"
+```
+
+**Regular Expression:**
+```json
+{"t": "/[a-z]+/"}
+```
+
+Note: Regular expressions in JSON must be represented as objects with a `t` property containing the regex pattern as a string.
+
+**Array of Alternatives:**
+```json
+{"t": ["true", "false"]}
+```
+
+**With Repetition:**
+```json
+{"t": [" ", "\t", "\n"], "repeat": "*"}
+```
+
+**Full Terminal Node:**
+```json
+{
+  "t": "/\\d+/",
+  "type": "Number",
+  "sample": "123",
+  "ast": ["num", ["$", "/raw"]]
+}
+```
+
+#### Important Notes
+
+**Regular Expression Syntax:**
+- In TypeScript/JavaScript, regex can be written as `/pattern/flags`
+- In JSON, regex must be a string: `{"t": "/pattern/flags"}`
+- Escape characters must be double-escaped in JSON strings: `"\\d+"` instead of `\d+`
+
+**Repetition Patterns:**
+- `repeat: "*"` means zero or more matches (equivalent to regex `*`)
+- `repeat: "+"` means one or more matches (equivalent to regex `+`)
+- Only applicable when `t` is an array of strings
+
+#### Examples
+
+```json
+{
+  "cst": {
+    "Null": "null",
+    "Number": "/\\-?\\d+(\\.\\d+)?/",
+    "Boolean": {"t": ["true", "false"]},
+    "Whitespace": {"t": [" ", "\t", "\n"], "repeat": "*"},
+    "Identifier": {
+      "t": "/[a-zA-Z_][a-zA-Z0-9_]*/",
+      "type": "Identifier",
+      "sample": "variable_name"
+    }
+  }
+}
+```
+
+### 3. ProductionNode (Production Node)
```

Comment:
Show some example of production node with value example that it matches against.